Vessel refactor

dseidman opened the pull request:
Created Vessel class to centralize and simplify all vessel-centric operations (temperature and volume). This commit only creates the class but doesn't use it anywhere, to allow review before ripping out the old code. It also inherits a few existing hacks that would have expanded scope too much to rip out (global temp[] array for aux temperatures and cycleTime[] timer array, OneWire bitrate #defines). It currently lacks PWM, steam, and temperature max from the existing code - those will be added back later. It does add a small amount of memory bloat, but only a few bytes, and should have a minimal runtime performance change. The class is designed to be deleted and recreated when config (eeprom) information changes, rather than dynamically updating the existing instance of the object.

Please provide feedback!

dseidman committed:
Added customizable brewer name on splash screen

dseidman committed:
New Vessel class simplifies temp and vol management

dseidman committed:
Fixed tons of bugs, typos, references, etc. Still doesn't compile.

dseidman committed:
Minor changes

dseidman committed:
Merge branch 'feature/vessel-refactor' of https://github.com/dseidman/BrewTroller-Official into feature/vessel-refactor

dseidman committed:
Fixing more typos, changed all temperature functions to double for compat with PID library

dseidman committed:
Fixed typo in class name

dseidman committed:
Merge branch 'feature/vessel-refactor' of https://github.com/dseidman/BrewTroller-Official into feature/vessel-refactor

Conflicts:
	src/Vessel.h

dseidman committed:
Missed Vessel.h in previous commit

dseidman committed:
Merge branch 'feature/vessel-refactor' of https://github.com/dseidman/BrewTroller-Official into feature/vessel-refactor

dseidman committed:
Fixed the last compile errors

dseidman committed:
Commented out custom brewery name for now

dseidman committed:
Fixed bug using free instead of delete

dseidman commented:
In case it isn't clear - the custom brewery name has been commented out of this branch. This branch only contains the vessel changes.

ericyanush commented on src/Vessel.cpp:
```diff
@@ -0,0 +1,285 @@
+		//Temperature Sensors: HLT (0-7), MASH (8-15), KETTLE (16-23), H2OIN (24-31), H2OOUT (32-39),
+		//          BEEROUT (40-47), AUX1 (48-55), AUX2 (56-63), AUX3 (64-71)
+		//**********************************************************************************
+		EEPROMreadBytes(8 * eepromIndex, sensorAddress, 8);
```
By moving the eeprom reading and writing directly into the vessel class we limit the ability to maintain an obvious and cohesive data storage schema in the eeprom. Any changes to the layout of the storage schema in the eeprom.h/cpp files have the potential to break the reads and writes here. 

Although I am not a fan of the current eeprom storage implementation we use, I think that maintaining the storage schema in the eeprom implementation and have each client use its interface makes changes to the data schema or implementation transparent to the client, and is the right way to prevent fragility in persistent data storage.

dseidman replied:
I agree. You mentioned that you had a good implementation of eeprom
encapsulation so I was planning to do all the eeprom interaction using
that, and figuring we would not make any eeprom schema changes until we had
better encapsulation. If you want to send me that implementation I'll
incorporate it into this project. Otherwise I can implement an
encapsulation class myself. Basically the vessel load code should be able
to call byte\* EEPROM->fetchVesselTemperatureSensorID(VS_HLT), bool
fetchVesselUsesPID(VS_HLT), etc. Or we could do it as EEPROM->load(VS_HLT,
TEMP_SENSOR), that might be better.

I do think that the vessel class should be able to directly load its
initialization values itself, rather than some other function loading them
from eeprom and passing them to the class, which introduces opportunities
for errors and requires us to keep the eeprom load code and vessel code in
sync. Letting the vessel fetch values from eeprom lets us change the eeprom
schema and the vessel class independently, as long as we don't need to
add/cut values from eeprom.

Send over that encapsulation example and I'll incorporate it, or let me
know if you want me to write that myself.

On Thu, Dec 17, 2015 at 12:47 PM, Eric Yanush notifications@github.com
wrote:

> In src/Vessel.cpp
> https://github.com/BrewTroller/BrewTroller-Official/pull/5#discussion_r47958786
> :
> 
> > -       maxPower = initMaxPower;
> > -       minTriggerPin = initMinTriggerPin;
> > -       includeAux[0] = initIncludeAux[0];
> > -       includeAux[1] = initIncludeAux[1];
> > -       includeAux[2] = initIncludeAux[2];
> >   +
> > -       usesAuxInputs = includeAux[0] + includeAux[1] + includeAux[2];
> >   +
> > -       //Load settings from eeprom
> >   +
> > -       //Load temperature sensor information - needs to be before PID setup so we have the temp probe addresses in place
> > -       //**********************************************************************************
> > -       //Temperature Sensors: HLT (0-7), MASH (8-15), KETTLE (16-23), H2OIN (24-31), H2OOUT (32-39),
> > -       //          BEEROUT (40-47), AUX1 (48-55), AUX2 (56-63), AUX3 (64-71)
> > -       //**********************************************************************************
> > -       EEPROMreadBytes(8 \* eepromIndex, sensorAddress, 8);
> 
> By moving the eeprom reading and writing directly into the vessel class we
> limit the ability to maintain an obvious and cohesive data storage schema
> in the eeprom. Any changes to the layout of the storage schema in the
> eeprom.h/cpp files have the potential to break the reads and writes here.
> 
> Although I am not a fan of the current eeprom storage implementation we
> use, I think that maintaining the storage schema in the eeprom
> implementation and have each client use its interface makes changes to the
> data schema or implementation transparent to the client, and is the right
> way to prevent fragility in persistent data storage.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/BrewTroller/BrewTroller-Official/pull/5/files#r47958786
> .

dseidman committed:
Integrated Vessel class into codebase. All vessel operations now run
through the class. Initial commit, not yet tested.

dseidman committed:
Fixing stupid bugs. Should compile now.

dseidman committed:
Fixing more stupid errors.

dseidman committed:
Fixing hopefully the last set of compile blocking bugs

dseidman committed:
Few more bugs. Not sure I got the CMake stuff right.

dseidman committed:
Last bug fixes except possibly CMakeLists.

dseidman committed:
CMake stuff giving me trouble, removed FEEDFORWRD for now. Compiles, doesn't link, easy fixes coming.

dseidman committed:
Fixed some legacy function calls. Added full preheat support.

dseidman committed:
Fixing bugs, moving to new computer

dseidman committed:
Fixed last few bugs - it compiles and links now!

dseidman committed:
Fixed feedforward.

dseidman committed:
Fixed a couple bugs related to SETPOINT_MULT

dseidman committed:
Added minimum and maximum temperature support to vessel class. Not yet supported as configurable, but they have sensible defaults.

dseidman committed:
Added steam support. Not tested very well yet.

dseidman committed:
Added utility function for better Celsius support (not yet integrated)

dseidman committed:
Merge branch 'feature/vessel-refactor' of https://github.com/dseidman/BrewTroller-Official into feature/vessel-refactor

dseidman committed:
Tons of changes including PID+PWM pumps, Flow Control, valve control
managed inline with output code, and more. Doesn't compile, pushing just
to back up.

dseidman committed:
Last few changes to support new model. Haven't compiled yet.

dseidman committed:
A bunch of fixes in BTNic code, as well as some readibility tweaks

dseidman committed:
Fixed all my typos. Builds and links now.

dseidman committed:
Minor code cleanup
